File: leetcode/1041. Robot Bounded in Circle/main.ex
defmodule Solution do
  @spec is_robot_bounded(instructions :: String.t) :: boolean
  def is_robot_bounded(instructions) do
    {x, y, dir} = Enum.reduce(String.graphemes(instructions), {0, 0, :north}, &follow_instruction(&1, &2))
    cond do
        x == 0 and y == 0 -> true
        dir == :north -> false
        true -> true
    end
  end

  def follow_instruction("G", state), do: forward(state)
  def follow_instruction("L", state), do: left(state)
  def follow_instruction("R", state), do: right(state)

  def forward({x, y, dir}) do
    case dir do
        :north -> {x, y + 1, dir}
        :south -> {x, y - 1, dir}
        :east -> {x + 1, y, dir}
        :west -> {x - 1, y, dir}
    end
  end

  def left({x, y, dir}) do
    case dir do
        :north -> {x, y, :west}
        :south -> {x, y, :east}
        :east -> {x, y, :north}
        :west -> {x, y, :south}
    end
  end

  def right({x, y, dir}) do
    case dir do
        :north -> {x, y, :east}
        :south -> {x, y, :west}
        :east -> {x, y, :south}
        :west -> {x, y, :north}
    end
  end

end

File: leetcode/1041. Robot Bounded in Circle/main.rs
#[derive(PartialEq)]
enum Dir {
    North,
    East,
    West,
    South
}

struct State {
    pub x: i32,
    pub y: i32,
    pub dir: Dir,
}

impl State {
    pub fn new() -> State {
        State {
            x: 0,
            y: 0,
            dir: Dir::North
        }
    }

    pub fn forward(self) -> State {
        match self.dir {
            Dir::North => State {
                y: self.y + 1,
                ..self
            },
            Dir::East => State {
                x: self.x + 1,
                ..self
            },
            Dir::West => State {
                x: self.x - 1,
                ..self
            },
            Dir::South => State {
                y: self.y - 1,
                ..self
            }
        }
    }

    pub fn left(self) -> State {
        State {
            dir: match self.dir {
                Dir::North => Dir::West,
                Dir::East => Dir::North,
                Dir::South => Dir::East,
                Dir::West => Dir::South,
            },
            ..self
        }
    }


    pub fn right(self) -> State {
        State {
            dir: match self.dir {
                Dir::North => Dir::East,
                Dir::East => Dir::South,
                Dir::South => Dir::West,
                Dir::West => Dir::North,
            },
            ..self
        }
    }
}

impl Solution {
    // robot is going to be bounded if:
    // 1. it returns to origin, no matter its final rotation
    // because all directions are symmetrical
    // 2. it is facing any direction other than north at the end
    // because if it's rotated by 90deg either way, it returns to its original position in 3 turns, and if it's rotated by 180deg, it returns to its original position in 1 turn
    pub fn is_robot_bounded(instructions: String) -> bool {
        let mut state = State::new();
        for instruction in instructions.chars() {
            state = match instruction {
                'G' => state.forward(),
                'L' => state.left(),
                'R' => state.right(),
                _ => state,
            };
        }
        if state.x == 0 && state.y == 0 {
            return true;
        }
        if state.dir == Dir::North {
            return false;
        }

        return true;
    }
}